rfcs/2020-10-06-3791-composing-components-pt-1.md

```diff
@@ -0,0 +1,87 @@
+it easy to create pre-built "chunks" of config that users could configure as
+normal components. These would be bundles of lower-level components wired
+together with adjusted default values for the specific use case.
+
+## Motivation
+
+We need a way to rapidly assemble Vector components that address specific use
+cases. This will allow us to improve ease of use without spending significant
+development time on each individual use case. It will allow us to focus
+development time on reuseable components without forcing users to do the work of
+assembling them from scratch.
+
+## Internal Proposal
+
+There are multiple levels at which we could implement this type of
+functionality:
+
+1. Manually implement new component as config facade over one existing component
+2. Manually implement new component as config facade over one source and one
+   parsing transform
+3. Manually implement new component as config expanding to arbitrary pipeline of
+   components
+4. Automatically derive new component from data describing arbitrary pipeline of
+   components
+
+We currently are at level (1), where we can do things like implement the Humio
+sink as a wrapper around the existing Splunk HEC sink.
+
+The next simplest is level (2). While it's not implemented yet, we do have
+existing plans to introduce the idea of a codec attached to sources. This would
+allow users to directly configure how to parse the incoming data as part of the
+source config itself. With that feature implemented, it would be relatively
+straightforward to do something similar to level (1) but expanding to both
+a source and an included codec.
+
+Level (3) becomes more complicated. We currently have a limited ability for
+transforms to expand to multiple transforms via `TransformConfig::expand`, and
+this could theoretically be generalized to include sources and sinks as well.
+The main problem is that this does not mesh well with the config traits as they
+currently exist and the API can be confusing. To do this properly would likely
+involve deeper changes to the config traits to better support this kind of
+staged building.
+
+Finally, layer (4) would allow defining these compositions via TOML instead of
+Rust code. This is somewhat similar to the idea of snippets that has been
+floated previously, but with a few key differences. The main one is that they
+would be built directly into Vector at compile time instead of loaded at
+runtime. This means they would need to be integrated into our build process and
+changing them would require recompiling Vector. They would also require
+a sufficiently general composition API to be exposed via TOML, which would be
+difficult to come up with for such a wide variety of potential pipelines. For
+these two reasons, I doubt that level (4) is worthwhile right now (this could
+change when/if we have more data-driven config definition in general).
+
+My proposal is that we initally focus on level (2) while collecting data on use
```

Comment:
I feel like we have cases of `3` already, in that some sinks are a `codec-ish -> batchedhttp -> http`.